hi-ml/src/health_ml/utils/serialization.py

```diff
@@ -0,0 +1,158 @@
+        self.model_config = model_config
+        self.git_repository = git_repository
+        self.git_commit_hash = git_commit_hash
+        self.dataset_name = dataset_name
+        self.azure_ml_workspace = azure_ml_workspace
+        self.azure_ml_run_id = azure_ml_run_id
+        self.text_tokenizer = text_tokenizer
+        self.image_pre_processing = image_pre_processing
+        self.image_dimensions = image_dimensions
+
+    def get_metadata_from_azureml(self) -> None:
+        """Reads information about the git repository and AzureML-related info from the AzureML run context.
+        If any of those information are already stored in the object, those have higher priority.
+        """
+        if not is_running_in_azure_ml():
+            return
+        if not self.azure_ml_workspace:
+            self.azure_ml_workspace = RUN_CONTEXT.experiment.workspace.name
+        if not self.azure_ml_run_id:
+            self.azure_ml_run_id = RUN_CONTEXT.id
+        properties: Dict = RUN_CONTEXT.get_properties()
+        if not self.git_repository:
+            self.git_repository = properties.get("azureml.git.repository_uri", "")
+        if not self.git_commit_hash:
+            self.git_commit_hash = properties.get("azureml.git.commit", "")
+
+    def state_dict(self, strict: bool = True) -> Dict[str, Any]:
+        """Creates a dictionary representation of the current object.
+
+        :param strict: The setting for the 'strict' flag in the call to torch.jit.trace.
+        """
+        def bytes_or_none(o: Any) -> Optional[bytes]:
+            return _dump_to_stream(o).getvalue() if o is not None else None
+
+        if self.model is None or self.model_example_input is None:
+            raise ValueError("To generate a state dict, the model and model_example_input must be present.")
+        try:
+            traced_model = torch.jit.trace(self.model, self.model_example_input, strict=strict)
+        except Exception as ex:
+            raise ValueError(f"Unable to convert the model to torchscript: {ex}")
+        jit_stream = BytesIO()
+        torch.jit.save(traced_model, jit_stream)
+        try:
+            config_yaml = object_to_yaml(self.model_config) if self.model_config is not None else None
+        except Exception as ex:
+            raise ValueError(f"Unable to convert the model configuration to YAML: {ex}")
+        return {
+            ModelInfo.MODEL: jit_stream.getvalue(),
+            ModelInfo.MODEL_EXAMPLE_INPUT: self.model_example_input,
+            ModelInfo.MODEL_CONFIG_YAML: config_yaml,
+            ModelInfo.GIT_REPOSITORY: self.git_repository,
+            ModelInfo.GIT_COMMIT_HASH: self.git_commit_hash,
+            ModelInfo.DATASET_NAME: self.dataset_name,
+            ModelInfo.AZURE_ML_WORKSPACE: self.azure_ml_workspace,
+            ModelInfo.AZURE_ML_RUN_ID: self.azure_ml_run_id,
+            ModelInfo.TEXT_TOKENIZER: bytes_or_none(self.text_tokenizer),
```

Comment:
What's the standard huggingface format? I'm leaning towards serializing the object itself because otherwise you also need to know that this is a huggingface tokenizer, etc, so more overhead. Unless you feel strongly about that, I'd leave that as-is for now.

Reply:
I feel we should stick to the huggingface way of saving it. Aside from being more familiar and a robust way of saving tokenizers, if you pickle the tokenizer things switching between the standard and fast implementations of huggingface tokenizers are no longer possible, while this is very easy to do if we save through the huggingface library.

Reply:
@Shruthi42, can you describe what you mean by the "huggingface way of saving"?

Reply:
Discussed offline. One comment: We are no longer pickling the tokenizer, removing this field will make it obvious we should be looking for a text tokenizer elsewhere - in this case in "other_info". Should we remove this field?

Reply:
Let's revisit that question when we have a better sense of how we are using the model serialization.